src/query/api/v1/handler/prometheus/common.go

```diff
@@ -239,3 +298,64 @@ func RenderTagCompletionResultsJSON(
 
 	return renderDefaultTagCompletionResultsJSON(w, results)
 }
+
+type tag struct {
+	name  string
+	value string
+}
+
+func writeTagsHelper(
+	jw *json.Writer,
+	completedTags []storage.CompletedTag,
+	tags []tag,
+) {
+	if len(completedTags) == 0 {
+		jw.BeginObject()
+		for _, tag := range tags {
+			jw.BeginObjectField(tag.name)
+			jw.WriteString(tag.value)
+		}
+
+		jw.EndObject()
+		return
+	}
+
+	firstResult := completedTags[0]
+	name := string(firstResult.Name)
+
+	for _, value := range firstResult.Values {
+		copiedTags := make([]tag, len(tags)+1)
+		copy(copiedTags, tags)
+		copiedTags[len(tags)] = tag{name: name, value: string(value)}
+		writeTagsHelper(jw, completedTags[1:], copiedTags)
+	}
+}
+
+func writeTags(
+	jw *json.Writer,
+	results []*storage.CompleteTagsResult,
+) {
+	for _, result := range results {
+		jw.BeginArray()
+		writeTagsHelper(jw, result.CompletedTags, []tag{})
```

Comment:
pass nil as last argument instead of empty slice

Reply:
Sounds good